### PR TITLE
Use git submodule for libpfm.

### PR DESCRIPTION
**Pull request by nemequ**
FWIW I'd rather just use the system copy; I'd be happy to submit a patch for that instead.

### CODE REVIEW

**Comment by travisdowns** [comment]
When you say "... and with a patched libpfm this [works on icc]" - is the patched version of libpfm the one that the submodule will point to (on git.sf...) after this change, or is the patch somewhere else?

**Comment by travisdowns** [comment]
@nemequ the default answer for "using system copies" is "amost always no" - especially for small, self-contained libraries like `libpfm4`. I much prefer to point a specific version in an accessible<sup>1</sup> git repo (first preference), or package a specific version directly in this repo (second preference, license permitting) than to ever rely on system copies. There are just so many advantages to that approach:

  1. Everything "just works" after a `clone` without the need to install additional libraries via your package manager (of which there a bunch of flavors) or however else you are supposed to install packages.
  2. You can point to exactly the version you want rather than relying on the usually exactly 1 version packaged by any given distribution (and it's on 1 version _per_ distro: different distros or releases will have different versions!).
  3. You can just patch the library and update your pointer for fixes, no wait years for your patch to be released by distributions (if you can even get upstream to accept your patch). If upstream doesn't accept your patch, you can use a fork for your patch.
  4. You just have one build system mechanism everywhere that doesn't rely on the vagaries of how system libraries are installed, how to use their headers, etc.
  5. It works on systems like Windows that doesn't have the same model of "locally installed libraries".

I really think it radically reduces maintenance and system-specific bugs in the long run, and indeed many languages have kind of adopted this model (you get the exact version of all dependencies). Yes, I know it's not the traditional model for shared libraries on UNIX, but you'd really need some big countervailing advantages for me to want to use system libraries. Something like "it saves disk space" or "the initial clone is shorted" are not really enough.

Two benefits of using the system library that could be enough of a reason would include:

  1. It avoids in-process conflicts with different versions of the same library loaded by other components. This is a concern for a shared library, but not a standalone process like `uarch-bench` (but it could be a concern if one wanted to make an embeddable version of uarch bench).
  2. The library is tightly tied to the system/kernel/distribution and so will vary among systems. I.e., something the c library falls into that category (and yeah I'm not masochistic enough to try to include my own version of that), and probably some other ones that rely on OS-specific or hardware-specific stuff. 

---

<sup>1</sup> The "accessible" part is important. Something like github definitely qualifies as accessible. A person git repo hosted on a private server by an individual that might stop being available at any point probably doesn't, with a large grey area in between. If the repo isn't likely to be accessible in the future, then the preference swaps to an embedded copy.

**Comment by nemequ** [comment]
No, the submodule points to [the main repo for libpfm](https://sourceforge.net/p/perfmon2/libpfm4/ci/master/tree/). I've submitted [a PR](https://sourceforge.net/p/perfmon2/libpfm4/merge-requests/7/) with the changes necessary for ICC, but the project seems pretty minimally maintained so it'll probably be several months before it's merged.

Sorry, I could have been clearer on that point.

File: .gitmodules [hunk 1]
```diff
@@ -7,3 +7,6 @@
 [submodule "backward-cpp"]
 	path = backward-cpp
 	url = https://github.com/bombela/backward-cpp.git
+[submodule "libpfm4"]
+	path = libpfm4
+	url = https://git.code.sf.net/p/perfmon2/libpfm4
```
Comment (travisdowns) [inline .gitmodules]: As far as you know, this is the git view of the official `libpfm4` repository which lives on sourceforce?

Comment (travisdowns) [reply .gitmodules]: My comment here raced with your comment, it seems this answer is "Yes".

**Comment by travisdowns** [comment]
@nemequ - OK, so currently it wont' build on icc due to issue you linked. If you want I'm fine also to point to a fork of libpfm4 that fixes that issue.

Out of curiosity: does `icc` compile it OK if you include it from a C file rather than a C++ file? In that case it could be worked around in uarch-bench by using a C file for that functionality.

File: .gitmodules (hunk 1 above)
Comment (nemequ) [reply .gitmodules]: Yep. [Perfmon2's web site](http://perfmon2.sourceforge.net/) links to [right to it](https://sourceforge.net/p/perfmon2/libpfm4/ci/master/tree/).

**Comment by nemequ** [comment]
> @nemequ - OK, so currently it wont' build on icc due to issue you linked. If you want I'm fine also to point to a fork of libpfm4 that fixes that issue.

It's up to you whether you want to use the official repo or my fork; I have a pretty strong preference for using the official repos whenever possible, but I guess you don't really share that.

> Out of curiosity: does icc compile it OK if you include it from a C file rather than a C++ file? In that case it could be worked around in uarch-bench by using a C file for that functionality.

The problem for ICC is libpfm itself, not uarch-bench. You can actually compile libpfm with gcc (or probably clang) then compile uarch-bench with ICC (*i.e.*, `cd libpfm4 && make && cd .. && make CC=icc CXX=icpc`) and it will work. Or, of course, you could just use the system copy (😉) and everything would work as expected.

**Comment by travisdowns** [comment]
> It's up to you whether you want to use the official repo or my fork; I have a pretty strong preference for using the official repos whenever possible, but I guess you don't really share that.

I mean sure, I prefer the official repo - but if you have a required fix and it isn't accepted by the upstream repo, ever, what do you do? Another option that kind of falls in between is to use the official repo, but include your patch as some type of patch file in the consuming project (or another repo) and apply your patch on top of it at build time. Bleh.

When the official repo is a barely maintained and on SF, while the fork is on github and perhaps has more recent activity than the official one, then you can add a few chalk marks to the column for the fork, too (that may be the case here).

> It's up to you whether you want to use the official repo or my fork

Actually you can choose - since this affects just you at the moment (that I know of). Of course, I might have to change this later, e.g., if your fork is abandoned, but I'm fine with either approach now.

> The problem for ICC is libpfm itself, not uarch-bench.

Right, I see. Well yeah that's another case where the system library is useful: when the library is less portable along some axis (in this case, compiler support) than the consuming project, then at least you avoid having to compile it with all the compilers you support.

I would accept a patch that optionally uses the `libpfm4` from the local system, based on a config variable, but which leaves the default as using the cloned version. Would that work for you?

**Comment by nemequ** [comment]
> Actually you can choose - since this affects just you at the moment (that I know of). Of course, I might have to change this later, e.g., if your fork is abandoned, but I'm fine with either approach now.

I'm fine with using the official repo; it's trivial for anyone who needs it to pull in the patch (`(cd libpfm4 && git pull https://git.code.sf.net/u/nemequ/perfmon2 master)`), and I don't have to keep my repo up to date ;).

> I would accept a patch that optionally uses the libpfm4 from the local system, based on a config variable, but which leaves the default as using the cloned version. Would that work for you?

I agree that would be nice, but I don't think it's worth the maintenance burden. Besides, due to the structure of uarch-bench I don't think it's a big deal to use embedded copies of libraries… if you were shipping a shared library it would be different, especially if you wanted to be able to get it into distros, but it seems like you expect people writing benchmarks to extend uarch-bench.

For what it's worth, I think it would be awesome to turn uarch-bench into a library, or maybe an executable which dlopens a user-specified library. That way all people would need to do is implement a standard interface and uarch-bench could do the rest. Kind of like how libfuzzer works, sans the compiler integration of course.

**Comment by travisdowns** [comment]
> For what it's worth, I think it would be awesome to turn uarch-bench into a library, or maybe an executable which dlopens a user-specified library. That way all people would need to do is implement a standard interface and uarch-bench could do the rest. Kind of like how libfuzzer works, sans the compiler integration of course.

@nemequ  - yes, for sure. Originally I wanted to do a bunch of reverse-engineering of uarch features (that's why it's called uarch-bench), but much of the work has now been on a framework for precise benchmarks, which is probably useful apart from the body of uarch themselves. Pulling that stuff out into a library is probably useful since one would presumably often want to embed this in some larger application.

**Comment by travisdowns** [comment]
This introduced issue #38 - I think because the submodule reference itself was not committed (`.gitmodules` was but there is also a submodule commit that needs to happen), not sure if github dropped it or it wasn't in the original commit, etc. Probably it worked locally for you because you already pulled down the module.

**Pull request merged**